File: backend/internal/terminology/terminology.go
// Package terminology is the coded vocabulary a diagnosis or a complaint is recorded in
// (CP52, §8, §9.1, D-24).
//
// # What D-24 leaves open, and what it does not
//
// D-24 is about SNOMED CT: its use requires an Affiliate licence, and whether Bangladesh
// confers free use has to be verified before any SNOMED content is embedded. Nothing here is
// SNOMED-derived; `core.terminology_map` exists so that it can be layered on the day the
// answer comes back, and a standing invariant refuses any content from a system marked
// unusable — because "we remembered not to" is not a control.
//
// What is *not* open is ICD. WHO publishes it under permissive terms, and the plan's own
// recommendation is ICD as the backbone with an internal dictionary for complaints. The one
// thing still undecided is ICD-10 or ICD-11, and this package does not care: a system is a
// row, a version is a row, and every coding carries both.
//
// # Why a coding carries its version
//
// Criterion 2, and it is not bookkeeping. ICD-10's E11.9 and ICD-11's 5A11 are not the same
// concept, and a diagnosis recorded in 2026 has to still mean in 2032 what the person who
// recorded it meant. A code with no system and no version is a string.
package terminology

import (
	"context"
	"errors"
	"strings"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/AmlanWTK/DTHCMS/backend/internal/platform/dbgen"
)

// Store reads the terminology. There is no write path: a code set is loaded by migration and
// a clinic does not edit the WHO's classification.
type Store struct {
	pool *pgxpool.Pool
	q    *dbgen.Queries
}

func NewStore(pool *pgxpool.Pool) *Store {
	return &Store{pool: pool, q: dbgen.New(pool)}
}

// System is one terminology and what may be done with it.
type System struct {
	Code           string `json:"code"`
	TitleEN        string `json:"title_en"`
	TitleBN        string `json:"title_bn"`
	Publisher      string `json:"publisher"`
	LicenceNote    string `json:"licence_note,omitempty"`
	Usable         bool   `json:"usable"`
	DefaultVersion string `json:"default_version,omitempty"`
}

// Concept is one coded thing: a diagnosis, or a complaint in the clinic's own dictionary.
type Concept struct {
	System  string `json:"system"`
	Version string `json:"version"`
	Code    string `json:"code"`

	DisplayEN string `json:"display_en"`
	DisplayBN string `json:"display_bn,omitempty"`

	// The grouping a picker files this under, in both languages. Bengali diagnoses filed
	// under English chapter names is what half-bilingual looks like on a screen, and it
	// reads as an interface somebody translated the easy parts of.
	Heading   string `json:"heading,omitempty"`
	HeadingBN string `json:"heading_bn,omitempty"`

	// FavouriteRank is the clinic's own ordering, where it has one. Criterion 1 — the twenty
	// most common diagnoses in three keystrokes — is reached by knowing which twenty, not by
	// a cleverer search.
	FavouriteRank *int `json:"favourite_rank,omitempty"`

	// Tier and Score are why this result came where it did. Returned rather than hidden:
	// "why is that third" is the question every search gets asked, and a ranking nobody can
	// inspect is a ranking nobody can tune.
	Tier  int     `json:"tier,omitempty"`
	Score float64 `json:"score,omitempty"`
}

// Mapping is one concept's equivalent in another system.
type Mapping struct {
	ToSystem    string `json:"to_system"`
	ToVersion   string `json:"to_version"`
	ToCode      string `json:"to_code"`
	Equivalence string `json:"equivalence"`
}

// ErrUnknownConcept is a code that is not in the system and version named.
var ErrUnknownConcept = errors.New("terminology: no such concept")

// ErrUnusableSystem is a search against a terminology this deployment may not use.
//
// Refused rather than returning nothing, because the two mean different things: "no results"
// sends a clinician looking for a better spelling, and "we may not use SNOMED here" sends
// somebody to D-24.
var ErrUnusableSystem = errors.New("terminology: that terminology may not be used here")

// ErrUnknownSystem is a terminology nobody has registered. Distinct from ErrUnusableSystem,
// because "we have never heard of that" and "we are not licensed for that" send the person
// reading the message to two different places.
var ErrUnknownSystem = errors.New("terminology: no such terminology")

// ErrUnknownVersion is a version that system has never published here.
//
// Refusing rather than falling back to the default is the whole of criterion 2: a caller who
// asked for ICD-10 2019 and silently got 2016 has recorded a coding whose version is a lie,
// and the lie is only discovered years later by somebody trying to read it back.
var ErrUnknownVersion = errors.New("terminology: no such version of that terminology")

// ErrNoDefaultVersion is a system that is registered but carries no content yet — ICD-11,
// today. A caller may name a version explicitly the moment one is loaded.
var ErrNoDefaultVersion = errors.New("terminology: that terminology has no default version here")

// MaxResults bounds a search. A picker showing more than this is a picker nobody reads to the
// bottom of, and the answer to a query with three hundred matches is a better query.
const MaxResults = 25

// Systems lists the terminologies and what may be done with each.
func (s *Store) Systems(ctx context.Context) ([]System, error) {
	rows, err := s.q.CodeSystems(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]System, 0, len(rows))
	for _, row := range rows {
		system := System{
			Code: row.Code, TitleEN: row.TitleEn, TitleBN: row.TitleBn,
			Publisher: row.Publisher, LicenceNote: row.LicenceNote, Usable: row.Usable,
		}
		if row.DefaultVersion != nil {
			system.DefaultVersion = *row.DefaultVersion
		}
		out = append(out, system)
	}
	return out, nil
}

// Search finds concepts. The ranking is the query's, in one statement; this only bounds the
// request and refuses a terminology that may not be used.
func (s *Store) Search(ctx context.Context, system, version, query string, limit int) ([]Concept, error) {
	version, err := s.Resolve(ctx, system, version)
	if err != nil {
		return nil, err
	}
	if limit <= 0 || limit > MaxResults {
		limit = MaxResults
	}
	// An empty query is the favourites, not an error and not everything. A picker opens
	// before anybody has typed, and what it should show then is the clinic's own list.
	if strings.TrimSpace(query) == "" {
		return s.Favourites(ctx, system, version)
	}
	rows, err := s.q.SearchTerminology(ctx, dbgen.SearchTerminologyParams{
		PQuery: query, PSystem: system, PVersion: version, PLimit: int32(limit),
	})
	if err != nil {
		return nil, err
	}
	out := make([]Concept, 0, len(rows))
	for _, row := range rows {
		concept := Concept{
			System: row.System, Version: row.Version, Code: row.Code,
			DisplayEN: row.DisplayEn, DisplayBN: row.DisplayBn,
			Heading: row.Heading, HeadingBN: row.HeadingBn,
			Tier: int(row.Tier), Score: row.Score,
		}
		if row.FavouriteRank != nil {
			rank := int(*row.FavouriteRank)
			concept.FavouriteRank = &rank
		}
		out = append(out, concept)
	}
	return out, nil
}

// Favourites is the clinic's own list, in the order it was ranked.
func (s *Store) Favourites(ctx context.Context, system, version string) ([]Concept, error) {
	version, err := s.Resolve(ctx, system, version)
	if err != nil {
		return nil, err
	}
	rows, err := s.q.TerminologyFavourites(ctx, dbgen.TerminologyFavouritesParams{
		System: system, Version: version,
	})
	if err != nil {
		return nil, err
	}
	out := make([]Concept, 0, len(rows))
	for _, row := range rows {
		concept := Concept{
			System: row.System, Version: row.Version, Code: row.Code,
			DisplayEN: row.DisplayEn, DisplayBN: row.DisplayBn,
			Heading: row.Heading, HeadingBN: row.HeadingBn,
		}
		if row.FavouriteRank != nil {
			rank := int(*row.FavouriteRank)
			concept.FavouriteRank = &rank
		}
		out = append(out, concept)
	}
	return out, nil
}

// Concept reads one, so a screen can render a coding recorded years ago under a version
// nobody uses any more.
func (s *Store) Concept(ctx context.Context, system, version, code string) (Concept, error) {
	version, err := s.Resolve(ctx, system, version)
	if err != nil {
		return Concept{}, err
	}
	row, err := s.q.TerminologyConcept(ctx, dbgen.TerminologyConceptParams{
		System: system, Version: version, Code: code,
	})
	if errors.Is(err, pgx.ErrNoRows) {
		return Concept{}, ErrUnknownConcept
	}
	if err != nil {
		return Concept{}, err
	}
	concept := Concept{
		System: row.System, Version: row.Version, Code: row.Code,
		DisplayEN: row.DisplayEn, DisplayBN: row.DisplayBn,
		Heading: row.Heading, HeadingBN: row.HeadingBn,
	}
	if row.FavouriteRank != nil {
		rank := int(*row.FavouriteRank)
		concept.FavouriteRank = &rank
	}
	return concept, nil
}

// Mappings is where a concept maps in another system. Empty until D-24 answers; the method
// exists so that the day it does, nothing above this line changes.
func (s *Store) Mappings(ctx context.Context, system, version, code string) ([]Mapping, error) {
	version, err := s.Resolve(ctx, system, version)
	if err != nil {
		return nil, err
	}
	rows, err := s.q.TerminologyMappings(ctx, dbgen.TerminologyMappingsParams{
		FromSystem: system, FromVersion: version, FromCode: code,
	})
	if err != nil {
		return nil, err
	}
	out := make([]Mapping, 0, len(rows))
	for _, row := range rows {
		out = append(out, Mapping{
			ToSystem: row.ToSystem, ToVersion: row.ToVersion, ToCode: row.ToCode,
			Equivalence: row.Equivalence,
		})
	}
	return out, nil
}

// Resolve turns a request's system and optional version into the pair every coding must
// carry, and refuses a terminology this deployment may not use.
//
// It is one round trip and it happens before every read, which is deliberate: the licence
// state and the version set are properties of the deployment, not of the caller, and a check
// the caller can skip is not a check. The version it returns is the one the rows will carry,
// so a client that never names a version still receives one with every result — which is how
// criterion 2 survives a picker whose author did not read it.
func (s *Store) Resolve(ctx context.Context, system, version string) (string, error) {
	var (
		usable       bool
		defaultVer   string
		versionKnown bool
	)
	err := s.pool.QueryRow(ctx, `
		SELECT s.usable,
		       coalesce((SELECT v.version FROM core.code_system_version v
		                  WHERE v.system = s.code AND v.is_default), ''),
		       EXISTS (SELECT 1 FROM core.code_system_version v
		                WHERE v.system = s.code AND v.version = $2)
		  FROM core.code_system s
		 WHERE s.code = $1`, system, version).Scan(&usable, &defaultVer, &versionKnown)
	if errors.Is(err, pgx.ErrNoRows) {
		return "", ErrUnknownSystem
	}
	if err != nil {
		return "", err
	}
	// Licence before existence of content: a system nobody may use should say so even when
	// it happens to be empty, or the operator fixes the wrong problem.
	if !usable {
		return "", ErrUnusableSystem
	}
	if strings.TrimSpace(version) == "" {
		if defaultVer == "" {
			return "", ErrNoDefaultVersion
		}
		return defaultVer, nil
	}
	if !versionKnown {
		return "", ErrUnknownVersion
	}
	return version, nil
}

File: web/src/features/terminology/api/terminology.ts
import { fieldMessages } from '@dthcms/api-client';
import type { ApiError, components } from '@dthcms/api-client';

import { api, unwrap } from '@/lib/api';
import type { Locale } from '@/lib/i18n/config';

/**
 * The coded terminology client (CP52, §4.6).
 *
 * Four reads, no writes. Nothing here is audited and nothing here is a patient — a search
 * for "diabetes" is a search for a word, and the ledger has better things to hold.
 *
 * Three facts about the server shape everything below, and each of them is a thing this
 * layer must not quietly improve on.
 *
 * **An empty `q` is the favourites.** Not everything, and not an error. A picker opens
 * before anybody has typed, and the honest thing to show then is the clinic's own twenty.
 * `listFavourites` exists as well as the empty search because the contract gives it its own
 * endpoint for exactly that reason: a screen that opens on a list should not have to send a
 * search to get one.
 *
 * **The version in the response is the resolved one.** A caller may name no version, and
 * the server answers with the one it actually used. `codingOf` therefore reads the version
 * off the concept, never off a prop the screen was configured with — see its note. This is
 * the whole of acceptance criterion 2, and it is the one thing in this file that would be
 * expensive to discover was wrong.
 *
 * **A refusal is a sentence, not a status.** Searching SNOMED answers 422 because this
 * deployment is not licensed for it pending D-24, and an unloaded version answers 422 rather
 * than being silently replaced by the default. Both arrive as an `ApiError` carrying a field
 * message, and `refusalText` is what turns that into the words on screen. "No results" and
 * "we are not licensed for that" send a clinician to two different places.
 */

export type CodeSystem = components['schemas']['CodeSystem'];
export type Concept = components['schemas']['Concept'];
export type ConceptList = components['schemas']['ConceptList'];
export type ConceptMapping = components['schemas']['ConceptMapping'];

/**
 * What a coding is: three fields, and not one fewer.
 *
 * `E11.9` on its own is a string. `E11.9` in ICD-10 2019 is a diagnosis. The picker hands
 * this shape to its caller rather than a bare code so that a screen storing a coding
 * physically cannot store half of one.
 */
export interface Coding {
  system: string;
  version: string;
  code: string;
}

/**
 * A coding with the words to show it by — what the picker hands its caller.
 *
 * Both displays travel, not the one the picker happened to be rendering. The screen that
 * stores this may be printed in the other language tomorrow, and re-reading the catalogue to
 * find out what `E11.9` is called in Bangla is a round trip nobody should need to make.
 */
export type ConceptSelection = Pick<
  Concept,
  'system' | 'version' | 'code' | 'display_en' | 'display_bn'
>;

/** The server's ceiling, mirrored so a caller can ask for the whole page without guessing. */
export const MAX_RESULTS = 25;

/** Which terminologies exist, who publishes them, and which this deployment may use. */
export async function listCodeSystems(): Promise<CodeSystem[]> {
  const body = await unwrap(api.GET('/v1/terminology/systems'));
  return body.systems;
}

export interface SearchRequest {
  system: string;
  /** Omit for the system's default. The response says which was used. */
  version?: string;
  /** What was typed. Empty returns the favourites. */
  q?: string;
  limit?: number;
}

/**
 * Autocomplete, tolerant of a misspelling.
 *
 * Returned in the server's order and never re-sorted here. The tiers are a ranking somebody
 * tuned against how this clinic actually types, and a client that re-ordered them would be a
 * second opinion nobody can inspect.
 */
export async function searchConcepts(request: SearchRequest): Promise<ConceptList> {
  const { system, version, q, limit } = request;
  return unwrap(
    api.GET('/v1/terminology/search', {
      params: {
        query: {
          system,
          ...(version === undefined ? {} : { version }),
          ...(q === undefined ? {} : { q }),
          ...(limit === undefined ? {} : { limit }),
        },
      },
    }),
  );
}

/** The clinic's own list, in the order somebody ranked it. What a picker opens on. */
export async function listFavourites(request: {
  system: string;
  version?: string;
}): Promise<ConceptList> {
  const { system, version } = request;
  return unwrap(
    api.GET('/v1/terminology/favourites', {
      params: { query: { system, ...(version === undefined ? {} : { version }) } },
    }),
  );
}

/** One concept and its mappings. `mappings` is empty everywhere until D-24 answers. */
export async function readConcept(request: {
  system: string;
  version?: string;
  code: string;
}): Promise<{ concept: Concept; mappings: ConceptMapping[] }> {
  const { system, version, code } = request;
  return unwrap(
    api.GET('/v1/terminology/concept', {
      params: { query: { system, code, ...(version === undefined ? {} : { version }) } },
    }),
  );
}

/**
 * The words this concept is read in, in the reader's language.
 *
 * Bangla when there is Bangla and the interface is in Bangla; English otherwise. The
 * fallback is deliberate and it goes only one way: a clinician reading Bangla is better
 * served by an English diagnosis than by a blank row they cannot select. `display_en` is
 * required by the contract, so the last resort below should never fire — it is here because
 * a picker row with no text in it is a row a screen reader announces as nothing at all, and
 * the code is at least something a person can look up.
 */
export function conceptLabel(concept: Concept, locale: Locale): string {
  if (locale === 'bn' && concept.display_bn) return concept.display_bn;
  if (concept.display_en) return concept.display_en;
  return concept.display_bn || concept.code;
}

/**
 * The grouping this row is filed under, in the reader's language.
 *
 * Same fallback as the display, and the reason it exists at all is a screenshot: the first
 * Bangla render of this picker showed a list of Bengali diagnoses filed under English
 * chapter names. Half-bilingual is its own failure — it reads as an interface somebody
 * translated the easy parts of. A standing database invariant now refuses a heading with no
 * Bangla form, so the empty case here should be unreachable for seeded content; it stays
 * because an empty heading is legal for a concept that has none.
 */
export function conceptHeading(concept: Concept, locale: Locale): string {
  if (locale === 'bn' && concept.heading_bn) return concept.heading_bn;
  return concept.heading ?? '';
}

/**
 * The three fields that make a coding, taken from the concept the server returned.
 *
 * Read off the concept rather than assembled from the picker's props, because those are two
 * different versions whenever the caller named none: the prop is absent and the server
 * resolved a default. A coding stamped with a version nobody searched is a lie that is only
 * discovered years later, when somebody asks what `E11.9` meant in 2019.
 */
export function codingOf(concept: Concept): Coding {
  return { system: concept.system, version: concept.version, code: concept.code };
}

/**
 * The concept, narrowed to what a caller stores.
 *
 * Built from the concept alone, for the same reason `codingOf` is: every field here is one
 * the server returned about this row, and none of them is a default the screen supplied.
 */
export function selectionOf(concept: Concept): ConceptSelection {
  return {
    ...codingOf(concept),
    display_en: concept.display_en,
    ...(concept.display_bn === undefined ? {} : { display_bn: concept.display_bn }),
  };
}

/** The message keys `tierReason` chooses between. Exported so a caller can exhaust them. */
export type TierReasonKey = 'reason.code' | 'reason.favourite' | 'reason.word' | 'reason.spelling';

/**
 * Why this row is where it is, as a message key.
 *
 * A ranking nobody can inspect is a ranking nobody can tune — the contract returns the tier
 * for that reason, and a picker that received it and drew nothing would be throwing away the
 * only feedback the person tuning it ever gets. It also answers the question a clinician
 * actually asks of a fuzzy match: *why is this here?* "Close to what you typed" is the
 * difference between trusting the fourth row and scrolling past it.
 *
 * `null` on the favourites, which carry no tier because they are not a search. Their rank is
 * shown instead, and it says more than a tier would.
 */
export function tierReason(concept: Concept): TierReasonKey | null {
  switch (concept.tier) {
    case 1:
      return 'reason.code';
    case 2:
      return 'reason.favourite';
    case 3:
      return 'reason.word';
    case 4:
      return 'reason.spelling';
    default:
      return null;
  }
}

/**
 * What the server said, rather than what the client guessed.
 *
 * A 422 here is never generic. It is "SNOMED CT is not licensed for this deployment" or
 * "this version has not been loaded", and both send the person somewhere specific — to the
 * licence decision, or to whoever loads content. Replacing either with "something went
 * wrong" turns a solvable problem into a support call.
 *
 * The field messages come first because they name the parameter that was wrong; the
 * envelope's own message is the fallback. Bangla where the server has it, English where it
 * does not, because silence is worse than the wrong language.
 */
export function refusalText(error: ApiError, locale: string): string | null {
  const named = Object.values(fieldMessages(error, locale)).filter((text) => text.trim() !== '');
  if (named.length > 0) return named.join(' ');

  const envelope = locale === 'bn' ? error.messageBN || error.messageEN : error.messageEN;
  return envelope.trim() === '' ? null : envelope;
}
